[CHORE] Refactoring of Types - Part 1: FOOD

AzaroQL opened the pull request:
# Description

Goals of these changes(will come in multiple PRs):
- create a single source of truth for ingame items
- simplify usability and extension of existing item objects
- cleanup redundant code
- force type safety to prevent potential bugs

This is part 1 of changes, refactored food(consumables) objects.
Main files:
src/features/game/types/consumables.ts
src/features/game/types/craftables.ts
src/features/game/types/items.ts (previously named images.ts)

Includes few minor fixes across the app.


## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

yarn test + manual test

# Checklist:

- [x] Title of the PR is relevant and is prefixed with [FEAT], [CHORE] or [FIX]
- [ ] Screenshot if it includes any UI changes
- [x] I have read the contributing guidelines and agree to the T&Cs
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

AzaroQL commented on src/features/game/expansion/lib/boosts.test.ts:
```diff
@@ -1,34 +0,0 @@
-import Decimal from "decimal.js-light";
```
Redundant tests, that logic is covered in src\features\game\events\landExpansion\fulfillGrubOrder.test.ts

Polysemouse commented:
The refactorings are much needed!  The item definitions are currently all over the places with different structure definitions

adamhannigan reviewed (commented):
Great initiative @AzaroQL 

The only suggestion I would make is to keep the consumable images inside of the original location.

Keeping our 'type' files purely focussed on types helps in a few areas:

- Avoid circular dependencies
- Run test suites easier (we don't need to worry about importing images in a non safe run time)

Everything else looks good to go!

AzaroQL commented:
My intention was to keep all info in one place since its easier to maintain, currently we have KNOWN_IDS(contains only item id), ITEM_DETAILS(contains image and description) and 'CRAFTABLES'(contains every other property and has duplicated description from ITEM_DETAILS)

I can create a method that will return object without image field to be used in tests.

AzaroQL commented:
Moved images to a separate file, created function that will fill images for ITEM_DETAILS object, if image is not available question mark will be used.
Refactored CROPS and CROP_SEEDS to objects from functions for type safety.

adamhannigan commented on src/features/game/types/crops.ts:
```diff
@@ -15,17 +17,16 @@ export type CropName =
  * Crops and their original prices
  */
-export const CROPS: () => Record<CropName, Crop> = () => ({
+export const CROPS: Record<CropName, Crop> = {
```
The func is needed in this case to ensure that we always get the current marketRate.

Every time we access `CROPS()` we need to calculate the cost at run time.

This means if someone is playing during a halvening event, the prices update automatically. With the func removed, these values will be set at initialisation time and a user will be interacting with the incorrect prices

adamhannigan commented on src/features/game/types/crops.ts:
```diff
@@ -92,11 +93,11 @@ export const CROPS: () => Record<CropName, Crop> = () => ({
 export type CropSeedName = `${CropName} Seed`;
 
-export const CROP_SEEDS: () => Record<CropSeedName, CraftableItem> = () => ({
+export const CROP_SEEDS: Record<CropSeedName, CraftableItem> = {
```
Same as above, the func is needed to ensure the correct values are always shown

adamhannigan commented on src/features/game/types/images.ts:
```diff
@@ -227,1523 +21,66 @@ import kaleMushroomPie from "assets/food/mushroom_kale_pie.png";
-    image: SUNNYSIDE.icons.expression_confused,
-    description: "?",
-  },
+export const IMAGES: Record<number, ItemDetailsWithImages> = {
```
What is the goal of indexing images by ID?

From a developer experience POV, we want to avoid using IDs where possible and use the string representation.

E.g.

`Pumpkin Soup`: pumpkinSoup.

This makes it much easier to read and follow throughout the code base

adamhannigan commented on src/features/game/types/consumables.ts:
```diff
@@ -33,163 +55,12 @@ export type ConsumableName =
-
+export const CAKES: Record<CakeName, Consumable> = {
   "Sunflower Cake": {
+    id: 505,
```
As per out chat, let's keep the IDs centralised in a single file.

adamhannigan commented on src/features/game/types/consumables.ts:
```diff
@@ -33,163 +55,12 @@ export type ConsumableName =
-    marketRate: 184,
-  },
-
+export const CAKES: Record<CakeName, Consumable> = {
```
Is there any relevance to `CAKES` anymore?

I would suggest calling these `BAKERY_CONSUMABLES` so it is clear the use case

adamhannigan reviewed (changes requested):
Getting better 💪 

Just noticed that the funcs were removed in a few places that are actually needed. Let me know if the comments make sense 👍

adamhannigan commented:
@AzaroQL This PR has sparked much discussion in our team lately, especially with the introduction of treasure island

We are facing similar architecture issues on our backend at the moment and need an overhaul.

At the moment can I ask that we pause working on this PR 🙏 We are using this PR as inspiration for the refactor. Right now the team will need to come up with a consistent approach we can adopt app wide and on FE and BE

AzaroQL commented:
Sure thing, feel free to close it if needed.
